### PR TITLE
Add --directory mode

### PR DESCRIPTION
**Pull request by bengourley**
Tools that don't end up bundle JS into a single output file can leave many source maps on the filesystem (e.g. Babel, TypeScript). The added `--directory` flag can be used to tell bugsnag-sourcemaps to search the filesystem from a given path and upload every source map that it
finds.

I made quite a substantial refactor to the library – splitting out the large `index.js` into smaller modules. Other than that – refactor-wise the scope of the changes were on a "needed to facilicate this new feature" basis.

In order to make the listr module output a reasonable amount of information to the CLI user, it would have necessitated tighter coupling between the uploader and the CLI and a further substantial refactor, so I removed that and implemented a simple logger instead. For this reason the CLI UI will appear very different to users. This may be a concern, but I felt it necessary for the bulk upload to show more than this amount of information:

image

Bulk upload outputs this kind of info:
image

Errors look like this:
image

## Testing

To facilitate manually testing the new mode I've added two kinds of Node project to the examples directory: TypeScript and Babel. As you'll see after running the build on each one, TypeScript creates a source map alongside each source/output file in the tree whereas Babel copies the `src` structure and mimics it in the output `lib`. 

### TypeScript

Be sure to replace the API key in `examples/typescript/package.json` and `examples/typescript/src/services/bugsnag.ts` before you begin.

```
cd examples/typescript
npm run build
npm run upload-source-maps
node src/app.js
```

### Babel

Be sure to replace the API key in `examples/babel/package.json` and `examples/babel/src/services/bugsnag.js` before you begin.

```
cd examples/babel
npm run build
npm run upload-source-maps
node lib/app.js
```

You can run the local developement `bugsnag-sourcemaps` command from either example using a relative path (per the package.json scripts): `../../cli.js upload --api-key --overwrite` etc.

With each project, you should be able to see errors in the dashboad being successfully symbolicated – showing the original stacktraces.

### CODE REVIEW

**Comment by Cawllec** [comment]
It's not entirely clear which changes are the `--directory` implementation, and which are refactoring.  Would it be possible to split this into two PRs, one for the refactor (or whichever change came first) and a second for the directory change?

**Comment by bengourley** [comment]
The commits are discreet chunks of work so you should be able to see from that? I can separate into a different PR if necessary.

I've updated the PR description adding directions on how to test the new feature.

File: README.md [hunk 1]
```diff
@@ -74,9 +92,9 @@ upload({
 });
 ```
 
-### Bugsnag Enterprise
+### Bugsnag On-premise
 
-If you are using Bugsnag Enterprise (on premise installation) with a custom domain, you can pass an optional `endpoint` option to define a custom upload url.
+If you are using Bugsnag On-Premise, you use the `endpoint` option to set the url to your upload server.
```
Comment (Cawllec) [inline README.md]: "you use the `endpoint`" => "you should use the `endpoint`"

**Review by Cawllec — COMMENTED** [review]
Ok, this looks good to me.  The majority of the changes are moving methods from `index.js` into discrete modules, with a few changes to use `graceful-fs` over built-in `fs`.  The `upload` -> `uploadOne` and `uploadMany` makes sense, as well as the options in the README and `cli.js` (which make up the `--directory` change.

I've also tested using the attached examples, and can confirm it works as described with both an empty `--directory` option and with an explicit path there (i.e. `--directory lib/`

I've left one comment on the readme, and I think it could do with a couple of things, namely:
- Unit tests for the `upload`, `uploadOne`, & `uploadMany` (stubbing `sendRequest`, how difficult would this be?)
- README's for the example apps.

File: README.md (hunk 1 above)
Comment (sazap10) [inline README.md]:
```suggestion
If you are using Bugsnag On-premise, you use the `endpoint` option to set the url to your upload server.
```
Should be On-premise

**Comment by bengourley** [comment]
> with a few changes to use graceful-fs over built-in fs

`graceful-fs` was in use before. If it's used in places it wasn't before, that's unintentional?

**Comment by Cawllec** [comment]
You are correct, I read it wrong initially.

**Comment by bengourley** [comment]
Added readmes for the examples. I do want to add unit tests but I wanted to get it queued up for review first. Will add tests on Monday.

**Review by Cawllec — APPROVED** [review]
Furthering my previous review, LGTM and tests pass :)
